# AI Machine Learning Practice 

# Informations

- Dataset: Music dataset

- Objectives: Classification

- Time Limits: 1 min

- Score: Classification Accuracy (Test Data)

- Please read all markdowns carefully 

- About Dataset: Music Style Data
    - 348 float type music features (frequency, tone, tempo, timbre...)
    - Label: Music Style
        - 1: Melancholy
        - 2: Romantic
        - 3: Rhythmical
    

## [Step 0] Importing Packages

You must specify all the packages you use in this practice in the cell below.



In [1099]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [1100]:
from __future__ import print_function
import os
data_path = ['/content/drive/MyDrive/Colab Notebooks/AI_data/ML/final']

from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score,LeaveOneOut
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegressionCV
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler, MinMaxScaler, MaxAbsScaler
import numpy as np

## [Step 1] Read Data

Train dataset is in the 'data' directory


In [1101]:
import pandas as pd

# Import the data using the file path
filepath = os.sep.join(data_path + ['music_train_data.csv'])
data = pd.read_csv(filepath)

In [1102]:
data.head(1)

f1        f2        f3        f4        f5        f6        f7  \
0 -0.166614  0.284691 -0.011022 -1.028812  0.101653  0.498247 -0.314566   

         f8        f9       f10  ...      f340      f341      f342      f343  \
0  1.208697 -1.503008 -1.457764  ...  2.136721 -1.193955  0.040614  1.127366   

       f344     f345      f346      f347      f348  answer  
0  0.741521 -0.70773  0.077748  0.832992 -1.291423       2  

[1 rows x 349 columns]

In [1103]:
print(data.shape)
print(data.dtypes)
data

(650, 349)
f1        float64
f2        float64
f3        float64
f4        float64
f5        float64
           ...   
f345      float64
f346      float64
f347      float64
f348      float64
answer      int64
Length: 349, dtype: object


f1        f2        f3        f4        f5        f6        f7  \
0   -0.166614  0.284691 -0.011022 -1.028812  0.101653  0.498247 -0.314566   
1   -0.583950  0.651829 -1.754836 -0.512545 -1.063596  1.434039 -1.404162   
2   -0.214595  0.353967  0.749125 -0.123296 -0.809340  0.042847 -0.688971   
3   -0.044177  0.151251  1.017266 -0.809429 -0.558230 -0.649047 -0.790528   
4    0.779536 -0.761447  0.138536  2.406821  0.735203 -0.058044  0.358249   
..        ...       ...       ...       ...       ...       ...       ...   
645 -0.092240  0.244564  2.173754  0.034836  0.180263 -2.834412 -0.392688   
646  0.915814 -0.940471 -0.717929  0.888751  0.787062  0.897648  1.208942   
647  0.935501 -0.970338 -0.388963 -0.555268  0.342415  0.414480  0.010077   
648  0.299071 -0.186426 -0.949760 -1.151156 -0.052513  0.516760  0.909057   
649 -1.808685  1.647698  0.714630 -0.383750 -2.896258  0.363108  0.557782   

           f8        f9       f10  ...      f340      f341      f342  \
0    1.208697 -1.503008 -1.457764  ...  2.136721 -1.193955  0.040614   
1   -0.745222  0.054440  0.153028  ... -0.803444 -0.327357  1.022505   
2   -0.741502  0.059323  0.995969  ...  0.423848 -0.200293 -0.265410   
3   -0.798548  1.471307  0.045307  ...  0.601954 -0.092321 -1.319499   
4    1.027844 -0.043828 -1.221248  ... -0.673082 -0.100535 -0.623208   
..        ...       ...       ...  ...       ...       ...       ...   
645  0.684052 -0.745517  0.918739  ... -0.796906  0.086825 -1.014114   
646 -0.863827  0.117906  0.355986  ... -0.746864 -0.600670 -1.552048   
647 -0.959610 -0.218170  0.607368  ... -0.768960 -1.382635  0.147514   
648 -0.147741 -0.781873 -0.224130  ...  0.103654  1.257128  0.111375   
649 -0.665291  0.035565  0.505658  ... -0.821553 -0.290509 -1.151889   

         f343      f344      f345      f346      f347      f348  answer  
0    1.127366  0.741521 -0.707730  0.077748  0.832992 -1.291423       2  
1   -1.083422 -0.714399 -1.407135  0.846917 -1.562645  0.365577       1  
2   -1.182842 -0.416810  0.234092 -1.061020 -1.595486 -0.867771       1  
3   -0.673082 -0.816716 -4.172430  1.093709  1.106629  0.476974       1  
4    1.571379  0.270780 -0.792427 -0.402796  1.349432  0.174976       2  
..        ...       ...       ...       ...       ...       ...     ...  
645  0.221252  0.200125 -0.153418  1.596107  0.641362  1.504907       2  
646 -0.825841 -0.372074  0.319805 -0.981512 -0.956493  0.741365       1  
647 -0.699404 -1.162207  1.279939  0.647133  0.191886  0.421133       3  
648 -1.456077 -1.084994 -0.250769 -1.140101 -0.592954  0.498912       1  
649 -0.344792 -0.159026 -0.595187  0.742030 -0.136604  1.485533       1  

[650 rows x 349 columns]

In [1104]:
skew = pd.DataFrame(data.skew())
skew.columns = ['skew']
skew['too_skewed'] = abs(skew['skew']) > 0.75
print(skew['too_skewed'].value_counts())
skew['too_skewed']
#f_i=0
#for i in range (0,len(features)):
#    if skew['too_skewed'][i]==True:
#        X_data=X_data.drop('f'+str(i+1),axis=1)

False    192
True     157
Name: too_skewed, dtype: int64


f1         True
f2        False
f3        False
f4         True
f5         True
          ...  
f345      False
f346       True
f347      False
f348      False
answer    False
Name: too_skewed, Length: 349, dtype: bool

## [Step 2] Data Preprocessing

* Preprocessing Code below 
* You must explain your method in this markdown
* (Important) You must define transfrom function for test data

In [1105]:
data= data[data['answer'] != 2]

data

f1        f2        f3        f4        f5        f6        f7  \
1   -0.583950  0.651829 -1.754836 -0.512545 -1.063596  1.434039 -1.404162   
2   -0.214595  0.353967  0.749125 -0.123296 -0.809340  0.042847 -0.688971   
3   -0.044177  0.151251  1.017266 -0.809429 -0.558230 -0.649047 -0.790528   
5   -1.249737  1.255367 -0.454869  0.620856 -0.978787  1.246140  0.475909   
6   -1.361175  1.329596  1.329538  0.509233 -0.104528  0.187242 -1.161702   
..        ...       ...       ...       ...       ...       ...       ...   
644 -0.771888  0.859637 -1.072970 -0.767326 -0.442038 -1.394626 -0.664999   
646  0.915814 -0.940471 -0.717929  0.888751  0.787062  0.897648  1.208942   
647  0.935501 -0.970338 -0.388963 -0.555268  0.342415  0.414480  0.010077   
648  0.299071 -0.186426 -0.949760 -1.151156 -0.052513  0.516760  0.909057   
649 -1.808685  1.647698  0.714630 -0.383750 -2.896258  0.363108  0.557782   

           f8        f9       f10  ...      f340      f341      f342  \
1   -0.745222  0.054440  0.153028  ... -0.803444 -0.327357  1.022505   
2   -0.741502  0.059323  0.995969  ...  0.423848 -0.200293 -0.265410   
3   -0.798548  1.471307  0.045307  ...  0.601954 -0.092321 -1.319499   
5   -1.057635 -0.094185  0.555478  ... -0.770683 -1.076433  0.609262   
6   -0.719656  0.416012  0.034384  ... -0.046721 -0.766499 -0.950612   
..        ...       ...       ...  ...       ...       ...       ...   
644  0.311025  2.021840  0.121895  ... -0.601092  1.270033  1.351185   
646 -0.863827  0.117906  0.355986  ... -0.746864 -0.600670 -1.552048   
647 -0.959610 -0.218170  0.607368  ... -0.768960 -1.382635  0.147514   
648 -0.147741 -0.781873 -0.224130  ...  0.103654  1.257128  0.111375   
649 -0.665291  0.035565  0.505658  ... -0.821553 -0.290509 -1.151889   

         f343      f344      f345      f346      f347      f348  answer  
1   -1.083422 -0.714399 -1.407135  0.846917 -1.562645  0.365577       1  
2   -1.182842 -0.416810  0.234092 -1.061020 -1.595486 -0.867771       1  
3   -0.673082 -0.816716 -4.172430  1.093709  1.106629  0.476974       1  
5   -0.775354 -0.245557  1.352815 -0.402796 -0.778304 -0.938712       3  
6   -1.407353  0.342610 -0.253246 -0.402796 -3.401938  1.155330       1  
..        ...       ...       ...       ...       ...       ...     ...  
644 -0.920947  0.680547  0.368328 -0.112046 -0.948313  1.782973       1  
646 -0.825841 -0.372074  0.319805 -0.981512 -0.956493  0.741365       1  
647 -0.699404 -1.162207  1.279939  0.647133  0.191886  0.421133       3  
648 -1.456077 -1.084994 -0.250769 -1.140101 -0.592954  0.498912       1  
649 -0.344792 -0.159026 -0.595187  0.742030 -0.136604  1.485533       1  

[410 rows x 349 columns]

split 후 실험 결과 confusion matrix를 살펴 본 결과 1과 2, 2와 3은 대체로 잘 분리해 내는 것을 확인하였으며, 정답이 3일때 1이라고 잘못 선택하는 경우가 많은 것을  확인했습니다.
따라서 1과 3을 더 잘 분리 할 수 있는 항목들 위주로 선택한다면 더 좋은 선택을 할 수 있을 것이라 생각해 2를 제외하고 1과 3만을 남긴 채 correlation을 파악하고 기준치를 넘는 feature을 선택하는 것을 고려하였습니다.

In [1106]:
features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']

In [1107]:
y_data.value_counts()

1    336
3     74
Name: answer, dtype: int64

In [1108]:
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)

현재 label의 비율이 동일하지 않기 때문에 이를 고려해 더 정확한 예측을 위해 train과 test에 각각 같은 비율로 label이 나누어 지도록 나누었습니다.

In [1109]:
y_train.value_counts()

1    235
3     52
Name: answer, dtype: int64

In [1110]:
fields = list(X_data.columns)
correlations_1_3 = abs(X_data[fields].corrwith(data['answer']))

In [1111]:
correlations_1_3

f1      0.088274
f2      0.076161
f3      0.123491
f4      0.054840
f5      0.017319
          ...   
f344    0.228034
f345    0.207666
f346    0.071933
f347    0.135626
f348    0.044163
Length: 348, dtype: float64

전체적으로 correlation이 크지 않아 0.10를 넘는 점을 기준점으로 삼았습니다.

In [1112]:
data = pd.read_csv(filepath)

features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']


fields = list(X_data.columns)
correlations_all = abs(X_data[fields].corrwith(data['answer']))

limit_1_3=0.1
limit_all=0.1
for i in range (0,len(correlations_all)):
    if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all:
        X_data=X_data.drop('f'+str(i+1),axis=1)

x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)


이후 전체 데이터(1,2,3 모두 고려)에서 correlation을 다시 체크해 최종 feature을 선정합니다.

In [1113]:
# Sample Code - Min Max Scaling
import warnings
warnings.filterwarnings('ignore', module='sklearn')
msc = StandardScaler()

X_data = pd.DataFrame(msc.fit_transform(X_data),  # this is an np.array, not a dataframe.
                    columns=X_data.columns)

X_data.head(5)

f8       f17       f18       f21       f25       f29       f31  \
0  1.144901  0.477720  1.379565  1.135770  1.297872  1.208146  0.856101   
1 -0.731226 -1.143826 -1.299640 -1.000853 -1.138750 -1.495138 -0.639008   
2 -0.727654 -1.083712 -1.305416 -0.666627 -0.574347 -0.729407 -0.647731   
3 -0.782429 -0.411961 -0.227924 -1.224343 -1.370331 -0.112162 -1.474771   
4  0.971249  2.345688  1.561796  1.775238  2.240499  1.341273  1.146911   

        f33       f35       f36  ...      f324      f325      f327      f329  \
0  0.017388 -1.071246 -1.112733  ...  0.750036 -1.275585  0.218185  1.223834   
1  0.457965  0.655994  0.275751  ...  0.215724 -1.278889 -0.563707 -0.092317   
2  0.369186  0.481671  1.500380  ...  1.363273  1.677990 -1.099559  0.909973   
3 -1.924987  1.670801  0.933710  ...  2.063661 -1.192991 -2.279091  0.557266   
4  1.272196 -1.359138 -1.910494  ... -1.783981  1.798578  1.454081  0.686972   

       f331      f332      f333      f337      f342      f347  
0  0.340386  0.083226 -1.049584  0.160745  0.030032  0.827437  
1  1.052749 -0.917872 -0.160513 -0.805383  0.990219 -1.599077  
2  1.797036  1.699528  0.495165 -0.233403 -0.269227 -1.632341  
3  2.467751 -1.710871  1.013045  0.467021 -1.300015  1.104601  
4 -1.313609  1.401523 -1.618960  1.352200 -0.619116  1.350533  

[5 rows x 99 columns]

In [1114]:
# transform function
# Do not change the function name
def transform_test(X_test_data):
    X_test_data = msc.transform(X_test_data)
    return X_test_data

## KNN

In [1115]:
# Mute the sklearn warning
import warnings
warnings.filterwarnings('ignore', module='sklearn')

#scaling = MinMaxScaler()
scaling = StandardScaler()
#scaling=MaxAbsScaler()

x_train = pd.DataFrame(scaling.fit_transform(x_train),columns=x_train.columns)
x_test = pd.DataFrame(scaling.fit_transform(x_test),columns=x_test.columns)

In [1116]:
# Function to calculate the % of values that were correctly predicted

def accuracy(real, predict):
    return sum(real == predict) / float(real.shape[0])

In [1117]:
# Sample code - kNN Classification
acc=0
selected_k=0
for k in range (1,51,2):
    model = KNeighborsClassifier(n_neighbors=k,p=1)#manhattan

    model = model.fit(x_train, y_train)

    y_train_pred=model.predict(x_train)    
    
    loo=LeaveOneOut()
    cv_N=10
    scores = cross_val_score(model, x_train, y_train, cv=cv_N)
    if acc<=np.mean(scores):
        acc=np.mean(scores)
        selected_k=k
model = KNeighborsClassifier(n_neighbors=selected_k,p=1)
model = model.fit(x_train, y_train)
y_test_pred = model.predict(x_test)

In [1118]:
print(accuracy(y_train, y_train_pred))
print(accuracy(y_test, y_test_pred))

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.8967032967032967
0.8871794871794871
[[232   3   0]
 [  8 159   1]
 [ 33   2  17]]
[[97  4  0]
 [ 2 70  0]
 [14  2  6]]


0.8967032967032967

0.8871794871794871

[[232   3   0]

 [  8 159   1]

 [ 33   2  17]]

[[97  4  0]

 [ 2 70  0]

 [14  2  6]]

In [1119]:
# Sample code - kNN Classification
acc=0
selected_k=0
for k in range (1,51,2):
    model = KNeighborsClassifier(n_neighbors=k,p=2)#eclidean

    model = model.fit(x_train, y_train)

    y_train_pred=model.predict(x_train)
    cv_N=10
    loo=LeaveOneOut()
    scores = cross_val_score(model, x_train, y_train, cv=cv_N)
    if acc<=np.mean(scores):
        acc=np.mean(scores)
        selected_k=k
model = KNeighborsClassifier(n_neighbors=selected_k,p=2)
model = model.fit(x_train, y_train)
y_test_pred = model.predict(x_test)

In [1120]:
print(accuracy(y_train, y_train_pred))
print(accuracy(y_test, y_test_pred))

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.8769230769230769
0.882051282051282
[[230   5   0]
 [  9 158   1]
 [ 39   2  11]]
[[95  5  1]
 [ 3 69  0]
 [12  2  8]]


0.8769230769230769

0.882051282051282

[[230   5   0]

 [  9 158   1]

 [ 39   2  11]]

[[95  5  1]

 [ 3 69  0]

 [12  2  8]]

기존에 주어진 코드를 사용한 knn 모델의 경우 0.923의 accuracy를 보여주는 것을 확인 할 수 있습니다.

하지만 이 결과는 train과 test 데이터를 구분하지 않았으며, 때문에 predict 결과가 좋게 나올 확률이 높을수밖에 없다 생각하였습니다.

train과 test를 분리하여 진행하였을때의 값은 0.871로 확인 가능합니다.

## DT

In [1121]:
features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']

In [1122]:
for i in range (0,len(correlations_all)):
     if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all:
         X_data=X_data.drop('f'+str(i+1),axis=1)

In [1123]:
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)
x_train

f8       f17       f18       f21       f25       f29       f31  \
57   0.123215 -0.437875 -0.827913  1.049783  1.434749  1.386821  0.945910   
208 -0.977854 -1.353342 -1.455652 -0.369477 -0.483983 -0.482379 -0.199092   
604 -0.418950 -1.321341 -1.687299 -0.191823 -0.152085 -0.042713 -0.030047   
553  0.015149 -0.087963  0.109663 -0.797464 -0.872219 -1.059957 -0.727321   
198  1.481545  1.964258  1.404730 -0.749478 -0.760875  0.912674 -0.900337   
..        ...       ...       ...       ...       ...       ...       ...   
149 -0.327736  0.246594  0.179294 -0.659355 -0.715771  0.052831 -0.212988   
230 -0.826988 -0.912295 -0.809514  1.368949  1.299397  1.362883  1.423078   
274 -0.524189 -0.620965 -0.304751 -0.490379 -0.546725 -0.990041  0.118915   
165 -0.857878 -0.403941 -0.563671 -1.543338 -1.654701  0.565116 -2.025719   
225  0.750517  1.351546  1.487777  0.762085  0.815309  1.015681  0.625471   

          f33       f35       f36  ...      f324      f325      f327  \
57   0.687164 -1.047854 -1.391268  ... -0.546793 -1.737293  0.487737   
208 -1.401159  0.026267 -0.266899  ...  1.406322  0.609859 -0.361021   
604  0.602415 -0.226243 -0.060895  ...  0.547116  1.105771  0.661960   
553  0.726622  0.645607  1.080007  ...  0.455506 -0.331798 -1.163654   
198  0.554859  0.967356  1.520058  ... -0.730432  0.692100  0.847304   
..        ...       ...       ...  ...       ...       ...       ...   
149 -0.806875  0.085672 -0.646972  ...  1.011805 -0.007769 -1.159634   
230  1.891100 -1.314046 -1.697911  ...  0.140475 -0.221595 -0.261745   
274 -0.692243 -0.112743 -0.543299  ...  1.016014 -1.759498 -0.493490   
165 -1.353323  2.796172 -0.439135  ... -0.116276 -0.309593 -0.698676   
225  0.566672 -0.546859  1.501190  ... -0.376184  1.539180 -0.613536   

         f329      f331      f332      f333      f337      f342      f347  
57   1.130113 -0.416600 -2.286095  0.397337 -0.973827 -0.162977  1.196793  
208  1.234436  1.339236  0.481069 -1.265701 -0.477674  2.225777  0.725945  
604  0.388927 -0.218995  1.342459  1.079788  0.950023 -0.930668 -0.205016  
553 -0.334123 -0.066353 -0.797872  0.776512  0.023666 -1.612040 -0.789250  
198 -0.227237 -0.561785  0.687318 -2.135249 -0.123597 -1.554601  1.402979  
..        ...       ...       ...       ...       ...       ...       ...  
149  0.026844  1.109043 -0.054772  0.574981 -0.604384  0.455213 -0.261864  
230 -0.241682  0.294411 -0.272141  1.797974  0.798666  0.096457  0.244317  
274 -0.761431  0.892695 -1.816982 -0.371526 -0.204326 -0.543990 -0.346613  
165  1.487406 -0.084712 -0.421773  0.691959 -1.885833  2.291433 -3.225434  
225 -0.903583 -0.377331  1.088693  0.135894  1.210501 -0.723567  1.033468  

[455 rows x 99 columns]

In [1124]:
# Mute the sklearn warning
import warnings
warnings.filterwarnings('ignore', module='sklearn')

#scaling = MinMaxScaler()
scaling = StandardScaler()
#scaling=MaxAbsScaler()

x_train = pd.DataFrame(scaling.fit_transform(x_train),columns=x_train.columns)
x_test = pd.DataFrame(scaling.fit_transform(x_test),columns=x_test.columns)

In [1125]:
model = DecisionTreeClassifier()

model.fit(x_train, y_train)


y_test_pred = model.predict(x_test)

print(accuracy(y_test, y_test_pred))

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)


0.7846153846153846
[[80  3 18]
 [ 5 65  2]
 [13  1  8]]


0.7846153846153846

[[81  3 17]

 [ 7 64  1]

 [13  1  8]]

In [1126]:
model = DecisionTreeClassifier(criterion="gini",max_depth=300) #gini_tree
model.fit(x_train, y_train)

y_test_pred = model.predict(x_test)

print(accuracy(y_test, y_test_pred))

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.7794871794871795
[[82  4 15]
 [ 6 63  3]
 [14  1  7]]


0.7743589743589744

[[79  4 18]

 [ 6 65  1]

 [14  1  7]]

In [1127]:
model = entropy_tree = DecisionTreeClassifier(criterion="entropy",max_depth=300) #entropy_tree
model.fit(x_train, y_train)

y_test_pred = model.predict(x_test)

print(accuracy(y_test, y_test_pred))

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.8358974358974359
[[83  6 12]
 [ 3 68  1]
 [ 8  2 12]]


0.8256410256410256

[[82  6 13]

 [ 3 68  1]

 [ 9  2 11]]

## Logistic

In [1128]:
features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']

In [1129]:
for i in range (0,len(correlations_all)):
    if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all:
       X_data=X_data.drop('f'+str(i+1),axis=1)

In [1130]:
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)

In [1131]:
# Mute the sklearn warning
import warnings
warnings.filterwarnings('ignore', module='sklearn')

#scaling = MinMaxScaler()
scaling = StandardScaler()
#scaling=MaxAbsScaler()

x_train = pd.DataFrame(scaling.fit_transform(x_train),columns=x_train.columns)
x_test = pd.DataFrame(scaling.fit_transform(x_test),columns=x_test.columns)

In [1132]:
# Standard logistic regression
model = LogisticRegression(C=.10, max_iter=400).fit(x_train, y_train)
y_train_pred=model.predict(x_train)
y_test_pred=model.predict(x_test)

print(accuracy(y_train, y_train_pred));
print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.9736263736263736
0.8564102564102564
[[233   0   2]
 [  1 167   0]
 [  9   0  43]]
[[87  2 12]
 [ 2 70  0]
 [10  2 10]]


0.9736263736263736

0.8564102564102564

[[233   0   2]

 [  1 167   0]

 [  9   0  43]]

[[87  2 12]

 [ 2 70  0]

 [10  2 10]]

In [1133]:
# L1 regularized logistic regression
model = LogisticRegressionCV(Cs=10, cv=10, penalty='l1', solver='liblinear').fit(x_train, y_train)

y_train_pred=model.predict(x_train)
y_test_pred=model.predict(x_test)

print(accuracy(y_train, y_train_pred));
print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.9736263736263736
0.8615384615384616
[[233   0   2]
 [  0 168   0]
 [ 10   0  42]]
[[90  3  8]
 [ 2 69  1]
 [11  2  9]]


0.9736263736263736

0.8564102564102564

[[233   0   2]

 [  0 168   0]

 [ 10   0  42]]
 
[[90  3  8]

 [ 3 68  1]

 [11  2  9]]

In [1134]:
# L2 regularized logistic regression
model = LogisticRegressionCV(Cs=10, cv=10, penalty='l2').fit(x_train, y_train)

y_train_pred=model.predict(x_train)
y_test_pred=model.predict(x_test)

print(accuracy(y_train, y_train_pred));
print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.9384615384615385
0.8923076923076924
[[229   6   0]
 [  2 166   0]
 [ 18   2  32]]
[[95  2  4]
 [ 2 70  0]
 [11  2  9]]


0.9384615384615385

0.8923076923076924

[[229   6   0]

 [  2 166   0]

 [ 18   2  32]]

[[95  2  4]

 [ 2 70  0]
 
 [11  2  9]]

## NB


In [1135]:
features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']
print(data.skew())

f1       -0.845729
f2        0.217869
f3        0.469548
f4        1.285284
f5       -1.129208
            ...   
f345     -0.219461
f346      1.049577
f347     -0.620625
f348     -0.458434
answer    0.715272
Length: 349, dtype: float64


In [1136]:
for i in range (0,len(correlations_all)):
    if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all or skew['too_skewed'][i]==True:
        X_data=X_data.drop('f'+str(i+1),axis=1)

x_train

f8       f17       f18       f21       f25       f29       f31  \
0    0.168098 -0.410412 -0.805395  0.983559  1.386468  1.415551  0.901221   
1   -0.984636 -1.315294 -1.419900 -0.388958 -0.500419 -0.531710 -0.223905   
2   -0.399507 -1.283663 -1.646663 -0.217154 -0.174029 -0.073683 -0.057795   
3    0.054961 -0.064545  0.112414 -0.802848 -0.882211 -1.133410 -0.742965   
4    1.590164  1.963948  1.380177 -0.756442 -0.772715  0.921603 -0.912977   
..        ...       ...       ...       ...       ...       ...       ...   
450 -0.304013  0.266144  0.180577 -0.669288 -0.728360  0.025851 -0.237559   
451 -0.826690 -0.879346 -0.787384  1.292214  1.253362  1.390614  1.370107   
452 -0.509684 -0.591384 -0.293263 -0.505877 -0.562120 -1.060573  0.088582   
453 -0.859031 -0.376870 -0.546725 -1.524157 -1.651706  0.559529 -2.018825   
454  0.824834  1.358320  1.461473  0.705337  0.777309  1.028911  0.586345   

          f33       f35       f36  ...      f324      f325      f327  \
0    0.685068 -1.036354 -1.353849  ... -0.496992 -1.772127  0.446016   
1   -1.363985  0.058271 -0.230269  ...  1.416750  0.601835 -0.399084   
2    0.601912 -0.199059 -0.024410  ...  0.574865  1.103412  0.619488   
3    0.723783  0.689435  1.115692  ...  0.485101 -0.350578 -1.198256   
4    0.555250  1.017326  1.555434  ... -0.676930  0.685015  0.804033   
..        ...       ...       ...  ...       ...       ...       ...   
450 -0.780876  0.118811 -0.610075  ...  1.030186 -0.022848 -1.194253   
451  1.866363 -1.307628 -1.660277  ...  0.176421 -0.239117 -0.300236   
452 -0.668400 -0.083392 -0.506475  ...  1.034310 -1.794586 -0.530981   
453 -1.317049  2.881054 -0.402384  ... -0.075154 -0.328119 -0.735283   
454  0.566841 -0.525795  1.536579  ... -0.329823  1.541771 -0.650510   

         f329      f331      f332      f333      f337      f342      f347  
0    1.098094 -0.374203 -2.356565  0.399625 -0.951960 -0.157209  1.195890  
1    1.202310  1.355082  0.466474 -1.242021 -0.482688  2.144333  0.722135  
2    0.357661 -0.179587  1.345257  1.073298  0.867658 -0.896872 -0.214576  
3   -0.364655 -0.029253 -0.838292  0.773923 -0.008510 -1.553368 -0.802419  
4   -0.257878 -0.517193  0.676887 -2.100385 -0.147794 -1.498026  1.403350  
..        ...       ...       ...       ...       ...       ...       ...  
450 -0.004055  1.128370 -0.080187  0.574985 -0.602533  0.438411 -0.271775  
451 -0.272308  0.326056 -0.301945  1.782247  0.724501  0.092753  0.237532  
452 -0.791529  0.915293 -1.877979 -0.359348 -0.224150 -0.524311 -0.357048  
453  1.455024 -0.047334 -0.454598  0.690458 -1.814554  2.207592 -3.253648  
454 -0.933537 -0.335529  1.086367  0.141546  1.114024 -0.697332  1.031556  

[455 rows x 99 columns]

NB에서 사용하기엔 skew된 data는 적절하지 못하므로, skew된 데이터는 drop합니다.

In [1137]:
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)

In [1138]:
# Mute the sklearn warning
import warnings
warnings.filterwarnings('ignore', module='sklearn')

scaling = MinMaxScaler()
#scaling = StandardScaler()
#scaling=MaxAbsScaler()

x_train = pd.DataFrame(scaling.fit_transform(x_train),columns=x_train.columns)
x_test = pd.DataFrame(scaling.fit_transform(x_test),columns=x_test.columns)

In [1139]:
model = GaussianNB();

model.fit(x_train, y_train);

y_train_pred = model.predict(x_train);
print(accuracy(y_train, y_train_pred));
y_test_pred = model.predict(x_test);
print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.9032967032967033
0.8358974358974359
[[213   3  19]
 [ 11 153   4]
 [  6   1  45]]
[[83  0 18]
 [ 6 65  1]
 [ 5  2 15]]


0.9032967032967033

0.8358974358974359

[[213   3  19]

 [ 11 153   4]

 [  6   1  45]]

[[83  0 18]

 [ 6 65  1]
 
 [ 5  2 15]]

In [1140]:
model = BernoulliNB();

model.fit(x_train, y_train);

y_train_pred = model.predict(x_train);

print(accuracy(y_train, y_train_pred));
y_test_pred = model.predict(x_test);

print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.545054945054945
0.517948717948718
[[233   0   2]
 [155  13   0]
 [ 50   0   2]]
[[93  5  3]
 [66  6  0]
 [20  0  2]]


0.545054945054945

0.517948717948718

[[233   0   2]

 [155  13   0]

 [ 50   0   2]]

[[93  5  3]

 [66  6  0]

 [20  0  2]]

In [1141]:
model = MultinomialNB();

model.fit(x_train, y_train);

y_train_pred = model.predict(x_train);

print(accuracy(y_train, y_train_pred));
y_test_pred = model.predict(x_test);

print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.8285714285714286
0.8512820512820513
[[232   3   0]
 [ 23 145   0]
 [ 45   7   0]]
[[101   0   0]
 [  7  65   0]
 [ 21   1   0]]


0.8285714285714286

0.8512820512820513

[[232   3   0]

 [ 23 145   0]

 [ 45   7   0]]

[[101   0   0]

 [  7  65   0]

 [ 21   1   0]]

##최종 모델 선정

In [1142]:
features = data.columns[:-1]
X_data = data[features]
y_data = data['answer']

In [1143]:
for i in range (0,len(correlations_all)):
    if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all:
        X_data=X_data.drop('f'+str(i+1),axis=1)

위에서 언급한 것과 같이 feature을 선택하는 부분으로, correlation이 0.1을 넘는 feature만 선택하였습니다.

In [1144]:
x_train,x_test,y_train,y_test=train_test_split(X_data,y_data,
                                               test_size=0.3,
                                               shuffle=True,
                                               random_state=42,
                                               stratify=y_data)

In [1145]:
#scaling = MinMaxScaler()
scaling = StandardScaler()
#scaling=MaxAbsScaler()

x_train = pd.DataFrame(scaling.fit_transform(x_train),columns=x_train.columns)
x_test = pd.DataFrame(scaling.fit_transform(x_test),columns=x_test.columns)


Scaler 선정 또한 값에 유의미한 변화를 보여주는 것을 확인할 수 있었고, 현재 StandardScaler를 선택할 때 가장 우수한 결과를 보여줌을 확인 할 수 있어 StandardScaler를 선택하였습니다.

In [1146]:
model = LogisticRegressionCV(Cs=10, cv=10, penalty='l2').fit(x_train, y_train)

model.fit(x_train, y_train);

전체적으로 정확도가 0.8~0.9 사이의 값을 유지하며, confusion matrix를 통해 오답이 발생하는 부분을 대략적으로 확인 가능하였습니다.

기존의 경우와 마찬가지로 1과 3을 구분하는데 있어 어려움을 겪는 것을 확인할 수 있었고, 이전에는 3을 1으로 많이 분류하였지만 이 부분을 조금이나마 개선 시켰습니다.

기존의 정확도만 높이기 위해선 KNN 방식도 고려될 수 있었으나, KNN의 경우엔 1과 3을 구분하는 능력이 logistic 방식을 사용하는 경우보다 더 떨어진다 판단하였고, 지금의 정확도는 3이 1과 2에 비해 적은 비중을 차지하고 있기 때문임을 고려하면 KNN의 정확도를 보여지는 값보다 더 낮게 판단하는게 적절하다 생각하였습니다.

반대로 GaussianNB의 경우엔 이전과는 달리 3을 1로 분류하는 경우가 더 적어졌지만, 반대로 1을 3으로 판단하는 경우가 더 증가하였으며, 때문에 전체적인 정확도가 낮아짐을 확인 할 수 있었습니다. 1이 가장 많은 비율을 차지하고 있기 때문에, 비율이 조정됨에 따라 GaussianNB가 더 적절한 선택지가 될 수도 있지만, 현재 주어진 데이터 상에서는 logisticRegressionCV를 채택하는 것이 가장 적절하다고 생각됐습니다.

HyperParameter의 경우엔 Cs와 cv를 다른 값으로 여러차례 수행하였으나, 이 값이 현재 10,10에서 가장 적합하다고 판단하였으며, 값을 더 증가시키더라도 값에 큰 향상효과를 기대하기는 어려웠으며, 또한 너무 큰 값을 설정한다면 오히려 연산시간이 증가하면서 model 생성시간이 너무 길어지는 것 또한 고려해야했습니다. 이러한 점을 고려하였을 때, 현재 cs와 cv는 둘다 10이 적당하였으며, 위의 실험에서 penalty를 p1과 p2로 각각 시험한 것 중 p2의 결과가 가장 우수하였습니다.

## Analysis 

* Analyze your model's result
* You may use additional metrics (F1 Score, Confusion matrix) or visualize your results using plots
* Hint : PCA plot will help you understand the dataset (Which class is the most challenging class to classify?)
* Hint : You may also compare different models to choose the best model among classifiers what we learned in this semester

In [1147]:
# Your code here
y_train_pred=model.predict(x_train)
y_test_pred=model.predict(x_test)

print(accuracy(y_train, y_train_pred));
print(accuracy(y_test, y_test_pred));

cf_train=confusion_matrix(y_train,y_train_pred)
print(cf_train)

cf_test=confusion_matrix(y_test,y_test_pred)
print(cf_test)

0.9384615384615385
0.8923076923076924
[[229   6   0]
 [  2 166   0]
 [ 18   2  32]]
[[95  2  4]
 [ 2 70  0]
 [11  2  9]]


0.9384615384615385

0.8923076923076924

[[229   6   0]

 [  2 166   0]

 [ 18   2  32]]

[[95  2  4]

 [ 2 70  0]

 [11  2  9]]

현재 확인 가능 하듯, 정확도 부분에서 0.89로 다른 알고리즘들에 비해 미세하게 더 우세한 성능을 보여주는 것을 확인할 수 있으며, 특히 2의 경우엔 많은 오차 없이 잘 판단하는 모습 또한 확인 가능합니다.

다만 앞에서 언급했듯, 1과 3을 구분하는 것이 다소 떨어지며, 이 부분에선 더 개선이 필요합니다.

앞에서 언급하였듯 KNN에 비해 1과 3을 더 잘 분리하는 것을 확인할 수 있고, gaussianNB보다 정확도가 더 높은 것을 확인 할 수 있습니다.

가능하다면 모델을 두번 사용해, 이 모델로 1,3과 2를 우선 구분하고, 이후 1과 3을 잘 구분하는 모델을 선택하여 한번 더 학습을 진행한다면 전체적으로 더 잘 훈련된 모델을 생성할 수 있을 것으로 판단하였으나, Test data 코드 부분에서 주어진 코드에 많은 변경이 필요하다 생각되었습니다. 때문에 현재 주어진 상황에서 가장 적절한 코드는 logisticRegressionCV라고 생각하였습니다.


# Test data

* TA will check your model's test data accuracy
* (Important) Do not change the code below

In [1148]:
filepath = os.sep.join(data_path + ['music_test_data.csv'])
t_data = pd.read_csv(filepath)
features = t_data.columns
X_t_data = t_data[features]


for i in range (0,len(correlations_all)):
    if correlations_1_3[i] < limit_1_3 or correlations_all[i]<limit_all:
        X_t_data=X_t_data.drop('f'+str(i+1),axis=1)

scaling=StandardScaler()
X_t_data= pd.DataFrame(scaling.fit_transform(X_t_data),columns=x_train.columns)
y_pred = model.predict(X_t_data)
np.savetxt('out.txt', y_pred, fmt='%d', delimiter='\n')
print(y_pred)

[2 2 2 1 1 1 2 1 1 2 2 1 2 1 2 1 2 1 1 2 2 1 2 2 2 1 1 2 2 2 2 2 2 2 1 2 1
 1 2 2 2 1 1 1 1 1 1 1 2 2 1 1 2 2 3 1 1 2 1 1 1 1 2 2 1 2 1 3 2 1 2 2 1 1
 1 1 1 2 1 2 2 1 2 1 1 2 2 2 1 1 2 1 1 1 2 1 2 1 1 3 1 1 1 1 1 1 2 2 1 2 1
 1 1 1 2 1 1 1 1 1 1 1 1 1 1 3 1 1 2 1 1 1 1 2 1 1 1 1 2 1 1 1 3 1 2 1 2 1
 3 2 1 1 1 1 1 1 3 1 1 1 2 1]


결과 : 2 2 2 1 1 1 2 1 1 2 2 1 2 1 2 1 2 1 1 2 2 1 2 2 2 1 1 2 2 2 2 2 2 2 1 2 1 1 2 2 2 1 1 1 1 1 1 1 2 2 1 1 2 2 3 1 1 2 1 1 1 1 2 2 1 2 1 3 2 1 2 2 1 1 1 1 1 2 1 2 2 1 2 1 1 2 2 2 1 1 2 1 1 1 2 1 2 1 1 3 1 1 1 1 1 1 2 2 1 2 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 3 1 1 2 1 1 1 1 2 1 1 1 1 2 1 1 1 3 1 2 1 2 1 3 2 1 1 1 1 1 1 3 1 1 1 2 1 가 나옴을 확인 할 수 있습니다.